walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,53 +15,122 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.LinkedList;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Stack;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     /**
-     * importantRanges - store time ranges that are important for at least one attendee in the request
-     * busyIntervals - eventually store merged important time ranges
-     * freeIntervals - used to collect intevals that are between busyIntervals   
+     * importantRanges - store time ranges that are important for at least one mandatory attendee in the request (doesn't apply to opitonal attendees)
+     * optionalImportantRanges - store time ranges that are important for mandatory as well as optional attendees 
+     * busyIntervals - eventualy store merged important time ranges
+     * optionalBusyRanges - same as above but for mandatory and optional attendees
+     * freeIntervals - used to collect intevals that are between busyIntervals
+     * freeRangesWithOptionals - same as above but for mandatory and optional attendees   
      */
     ArrayList<TimeRange> importantRanges = new ArrayList<TimeRange>();
-    Stack<TimeRange> busyIntervals = new Stack<TimeRange>();
-    ArrayList<TimeRange> freeIntervals = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> optionalImportantRanges = new ArrayList<TimeRange>();
+    Stack<TimeRange> busyRanges = new Stack<TimeRange>();
+    Stack<TimeRange> optionalBusyRanges = new Stack<TimeRange>();
+    ArrayList<TimeRange> freeRanges = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> freeRangesWithOptionals = new ArrayList<TimeRange>();
     
+    //list that helps in optimizing search for important events
+    LinkedList<Event> eventsList = new LinkedList<Event>(events);
+    LinkedList<Event> eventsLeft = eventsList;
+    //populates importantRanges and optionalImportantRanges with ranges that are important for mandatory attendees
     for(String attendee : request.getAttendees()){
-      for (Event event : events){
+      eventsList = eventsLeft;
+      eventsLeft = new LinkedList<Event>();
+      for (Event event : eventsList){
         if(event.getAttendees().contains(attendee)){
           importantRanges.add(event.getWhen());
+          optionalImportantRanges.add(event.getWhen());
+        } else{
+          eventsLeft.add(event);
         }
       }
     }
```

Comment:
IIUC you're trying to find events which involve at least one of the required attendees here (for the part below it's for one of the optional attendees). The reason you need `eventsLeft` and `eventsList` is because if an event doesn't involve the attendee you're checking in the current loop, you need to save it and check back because it may involve another attendee in a future loop.

Can't we iterate events in the outer loop while iterating attendees in the inner loop? This way we can get rid of `eventsList` and `eventsLeft`, and we can largely reduce the overall number of looping, if we just break the inner loop when we see the first attendee that's involved in the current event.

Reply:
Oh, yes, that's correct. Such approach creates much cleaner code. Thanks.